kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -128,10 +140,26 @@ public void produce(
             serializedValue,
             request.getTimestamp().orElse(Instant.now()));
 
-    return produceResult.thenApply(
-        result ->
-            toProduceResponse(
-                clusterId, topicName, keyFormat, keySchema, valueFormat, valueSchema, result));
+    return produceResult
+        .handle(
+            (result, ex) -> {
+              if (ex != null) {
+                recordErrorMetrics();
+                throw new CompletionException(ex);
+              }
+              return result;
+            })
+        .thenApply(
+            result -> {
+              ProduceResponse response =
+                  toProduceResponse(
+                      clusterId, topicName, keyFormat, keySchema, valueFormat, valueSchema, result);
+              double responseSize = getResponseSize(response.getKey(), response.getValue());
+              double latency =
+                  Duration.between(requestInstant, result.getCompletionTimestamp()).toMillis();
```

Comment:
Isn't the actual latency result here a `long`? I think we only deceive ourselves that we have better precision than what we actually have. I'd suggest using `long` or even `int`, and if some metrics API requires `double`, let's convert at the call-site.

P.S. This change would then propagate all the way down.

Reply:
fixed